Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift

```diff
@@ -0,0 +1,302 @@
+            )
+        }
+
+        mutating func newHTTP2MaxConcurrentStreamsReceived(_ connectionID: Connection.ID, newMaxStreams: Int) -> Action {
+            let (index, context) = self.connections.newHTTP2MaxConcurrentStreamsReceived(connectionID, newMaxStreams: newMaxStreams)
+            return self.nextActionForAvailableConnection(at: index, context: context)
+        }
+
+        mutating func http2ConnectionGoAwayReceived(_ connectionID: Connection.ID) -> Action {
+            _ = self.connections.goAwayReceived(connectionID)
+            // TODO: create new connection if we have open requests on the given event loop
+            return .none
+        }
+
+        mutating func http2ConnectionClosed(_ connectionID: Connection.ID) -> Action {
+            _ = self.connections.failConnection(connectionID)
+            // TODO: create new connection if we have open requests on the given event loop
+            return .none
+        }
+
+        mutating func http2ConnectionStreamClosed(_ connectionID: Connection.ID) -> Action {
+            let (index, context) = self.connections.releaseStream(connectionID)
+            return self.nextActionForAvailableConnection(at: index, context: context)
+        }
+
+        mutating func failedToCreateNewConnection(_ error: Error, connectionID: Connection.ID) -> Action {
+            self.failedConsecutiveConnectionAttempts += 1
+            self.lastConnectFailure = error
+
+            let eventLoop = self.connections.backoffNextConnectionAttempt(connectionID)
+            let backoff = self.calculateBackoff(failedAttempt: self.failedConsecutiveConnectionAttempts)
+            return .init(request: .none, connection: .scheduleBackoffTimer(connectionID, backoff: backoff, on: eventLoop))
+        }
+
+        mutating func connectionCreationBackoffDone(_ connectionID: Connection.ID) -> Action {
+            // The naming of `failConnection` is a little confusing here. All it does is moving the
+            // connection state from `.backingOff` to `.closed` here. It also returns the
+            // connection's index.
+            guard let (index, _) = self.connections.failConnection(connectionID) else {
+                preconditionFailure("Backing off a connection that is unknown to us?")
+            }
+            let (newConnectionID, eventLoop) = self.connections.createNewConnectionByReplacingClosedConnection(at: index)
+            return .init(request: .none, connection: .createConnection(newConnectionID, on: eventLoop))
+        }
+
+        mutating func timeoutRequest(_ requestID: Request.ID) -> Action {
+            // 1. check requests in queue
+            if let request = self.requests.remove(requestID) {
+                var error: Error = HTTPClientError.getConnectionFromPoolTimeout
+                if let lastError = self.lastConnectFailure {
+                    error = lastError
+                } else if !self.connections.hasConnectionThatCanOrWillBeAbleToExecuteRequests {
```

Comment:
We can't use that here. Here we should only check for `active` connections. If we have a starting connection, but no active we need to assume `connectTimeout`.